### DISCUSSION TITLE
Windows链接Catch2失败

### DISCUSSION THREAD

**Discussion by Extra-Creativity** [xmake-io/xmake#3301] [Q&A] [answered]
xmake.lua
```lua
add_requires("catch2")
add_packages("catch2")

target("main")
    set_kind("binary")
    add_links("Catch2Main", "Catch2")
    add_files("a.cpp")
```
其中a.cpp为：
```c++
#include <catch2/catch_test_macros.hpp>

TEST_CASE("This is a test name", "[this is test tag]")
{
	CHECK(6 * 9 == 42);
}
```
工具链是VS2019的默认链，生成的时候提示`error: LINK : fatal error LNK1561: 必须定义入口点`；但是`Catch2Main.lib`里本身是有main的；根据[Catch2 Issue](https://github.com/catchorg/Catch2/issues/1830)，上面这种方式应该没什么问题。这个程序我在Linux下用xmake可以正常链接，但是Windows下不行，想问一下什么原因？

**Comment by waruqi** [comment, marked as answer]
>     add_links("Catch2Main", "Catch2")

删了，直接 add_packages 就行了

> 但是Windows下不行，想问一下什么原因？

main 定义在 static lib 里，windows 需要主程序设置 ldflags   显式指定 console subsystem 。。你自己 google 找下 flags 设置下。

**Reply by Extra-Creativity** [reply]
删了连Linux都链接不通过..Catch2就是main在libCatchMain.a里，就不用写main了，必须add_links吧？

**Reply by Extra-Creativity** [reply]
哦，我看了一下，是libCatch2Main里需要用到libCatch2，所以链接失败了；我感觉这个仅用`add_packages`好像解决不了。

**Reply by waruqi** [reply]
> 哦，我看了一下，是libCatch2Main里需要用到libCatch2，所以链接失败了；我感觉这个仅用`add_packages`好像解决不了。

我看 packages 里面不是已经 -lCatch2Main -lCatch2 么。。为啥不行。。

```
"/usr/bin/xcrun -sdk macosx clang++" -o build/macosx/x86_64/release/test build/.objs/test/macosx/
x86_64/release/src/main.cpp.o -target x86_64-apple-macos12.6 -isysroot /Applications/Xcode.app/Co
ntents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.sdk -stdlib=libc++ -L/Users/
ruki/.xmake/packages/c/catch2/v3.2.1/f4738f6b5882478792f98cefc28617e1/lib -Wl,-x -lCatch2Main -lC
atch2 -lz
```

**Reply by Extra-Creativity** [reply]
我刚把很多东西都更新到最新版本，但是还是不行；verbose是这样的：
```
/usr/bin/gcc -c -m64 -o build/.objs/main/linux/x86_64/release/a.cpp.o a.cpp
checking for flags (-MMD -MF) ... ok
checking for flags (-fdiagnostics-color=always) ... ok
[ 50%]: linking.release main
/usr/bin/g++ -o build/linux/x86_64/release/main build/.objs/main/linux/x86_64/release/a.cpp.o -m64
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../x86_64-linux-gnu/Scrt1.o: in function `_start':
(.text+0x24): undefined reference to `main'
/usr/bin/ld: build/.objs/main/linux/x86_64/release/a.cpp.o: in functio `CATCH2_INTERNAL_TEST_0()':
a.cpp:(.text+0x31): undefined reference to `Catch::StringRef::StringRef(char const*)'
...
```
请问您刚才回复那条，也是说自己链接有问题吗？还是上面的程序可以正常链接？

**Reply by waruqi** [reply]
> /usr/bin/g++ -o build/linux/x86_64/release/main build/.objs/main/linux/x86_64/release/a.cpp.o -m64

没任何 catch2 包的 links linkdirs 。。你确定 add_package 了？ 到 main target了？

**Reply by waruqi** [reply]
add_requires("catch2", {system = false}) 你是不是系统里面装过？禁用下系统库试试

**Reply by Extra-Creativity** [reply]
可以了，麻烦您了

**Reply by Extra-Creativity** [reply]
补充一下，以备其他查找的人使用；Windows链接的问题可以在`target("xxx")`下加一个`add_ldflags("/SUBSYSTEM:CONSOLE")`来解决。
